### ISSUE TITLE
Expandable menu for mobile website

### ISSUE THREAD

**Issue by jo2** [checkstyle/checkstyle#8244]
As mentioned below, during the implementation of the mobile version of the checkstyle website there were thoughts about having the menu being expandable. Is this stil something wanted? If so, I can take a look at it.

"Let's think about expandable menu in separate issue"

_Originally posted by @romani in https://github.com/checkstyle/checkstyle/issues/7083#issuecomment-554383874_

Demo: https://jo2.github.io/checkstyle-menu-suggestion/

**Comment by romani** [comment]
@jo2, thanks a lot for helping us.
It would be awesome is make a demo of how it will work on single page before you update whole site.

Let me know if you need any help, your site update is awesome, I love it each day when I reply to contributors and users.

**Comment by jo2** [comment]
Will work on it and psuh an example to mit github pages when its ready.

**Comment by jo2** [comment]
At [https://jo2.github.io/checkstyle-menu-suggestion/](https://jo2.github.io/checkstyle-menu-suggestion/) I made a suggestion how it could look. Please tell me what you think.

**Comment by rnveach** [comment]
This is just me and my thoughts, but I dislike having a collapsed menu that takes a new page just to load the expansion and the expansion isn't even scrolled into view (that I saw) to help with the transition. I am not sure how well we can do this since the same links are used for real pages that can be loaded.

If we do go the way of keeping things collapsed, I think I would like to see "Checks" still expanded for desktops since it is an important section and when I load the documentation that is 99% of the time the section I go to.

**Comment by jo2** [comment]
Oh, i was so focused on mobile view that i didn't notice that these menu entries are collapsed in desktop view, too. I'll fix that. 

To fix the first aspect you mentioned, we can use menu entries without a page and add something like an overview page where the content from the original page is. I'll update the suggestion in the next fiew days.

**Comment by rnveach** [comment]
@romani Share your thoughts so we are on the same page.

**Comment by romani** [comment]
Desktop should not be affected.
General Font on desktop is very small, it is old issue ... , separate issue, but yes it is not good, I use 120-130% zoom all the time for our pages to read text and navigate easily.

I play periodically with your version of web site on phone, I have some concerns:
- with fact that some pages have child pages but there is no navigation to them inside page
- not sure if "menu" hide/show is useful and beneficial, as it not stick to top, easy scrolling to bottom - hide it, but clicking on far left corner if you hanging phone in right hand is hard.

I need to play a bit more to make my feedback more firm.

**Comment by jo2** [comment]
@romani how about we make the top line containing date and version sticky to the top in mobile version? This would also make the menu open / hide button sticky to the top.

**Comment by romani** [comment]
Sounds interesting, please do some hack to make demo of it in GitHub.io .
Current demo version doesn't work properly ( menu is not hiding).

**Comment by jo2** [comment]
I updated it at [https://jo2.github.io/checkstyle-menu-suggestion/](https://jo2.github.io/checkstyle-menu-suggestion/). Please let me know what you think.

**Comment by rnveach** [comment]
@jo2 When on the desktop and scrunch the window very small to simulate a mobile, the actual page contents go away completely and all I can see is the left menu and nothing else. I am on Firefox.

**Comment by jo2** [comment]
@rnveach can you please post a screenshot of this behaviour? I'm using Firefox, too, and either I don't understand what you mean or I can't reproduce it.

**Comment by romani** [comment]
Feedback: 
- It would be much better to have menu on right side
- It would be good to not collapse all menu items, because if menu is easy to hide, size of it is not a problem. But it is problem to navigate in tree of pages.

**Comment by jo2** [comment]
I've updated my suggestion on [https://jo2.github.io/checkstyle-menu-suggestion/](https://jo2.github.io/checkstyle-menu-suggestion/)

@rnveach I managed to reproduce the behaviour you've mentioned. It is an issue with sticky position of the menu. Sticky position was used for displaying the menu at the top of the current window no matter at which scroll position the page is. I now removed the sticky position so the menu is only displayed at the top of the page. You'd now have to scroll to the top to see the menu though.

@romani I implemented your suggestions about the hamburger icon and the menu on the right and removed the collapsible from the tree of pages.

**Comment by romani** [comment]
I have very weird behavior :) (chrome on android), hamburger is always jumping from one position to another position in 
Here it is in the middle 
![Screenshot_20200611-063548](https://user-images.githubusercontent.com/812984/84391714-33cdb580-abae-11ea-94f8-a1fe87b05e7e.png)
but when you operate with it, it may go to be first in this line or last or middle.
Please help to fix it.

additional problem: when I navigate to certain Check documentation, hamburger line is on top but - good, BUT I cannot click on to expand menu again

**Comment by jo2** [comment]
In regards to your second problem: I think [this example](https://jo2.github.io/checkstyle-menu-suggestion/config_design.html#InnerTypeLast) represents what you've mentioned. The menu opens (you can see it in the scroll position in the scrollbar) but the menu is at the top of the page (as mentioned [here](https://github.com/checkstyle/checkstyle/issues/8244#issuecomment-634729047).

In regards to your main problem, I've added a fix for that. Due to maven-site-plugin enforcing a specific order of the breadcrumb content, i have to manipulate the order using js to display the bruger on the right side. if this resorting function is triggered again (for example due to resizing) the elements are resorted again. I added a check if the sorting already is correct. [https://jo2.github.io/checkstyle-menu-suggestion/](https://jo2.github.io/checkstyle-menu-suggestion/)

**Comment by jo2** [comment]
I made previews for both alternatives now:
* [menu in relative position](https://jo2.github.io/checkstyle-menu-relative/)
* [menu sticky at the top of the screen](https://jo2.github.io/checkstyle-menu-sticky/)

Please let me know what you think.

**Comment by romani** [comment]
Unfortunately non of them I like, user experience is weird.

What if address it differently.
We had a problem when user navigate in menu but due to long menu user do not see that he actually on different pages - this was original problem.

So we can make for mobile mode menu to be colabsed to single row by default, with link to expand. User control himself how to show it. When use scrolled to middle or end and he wants to open menu , we show him  floating link "scroll to top" that overlay whole content of page.

**Comment by jo2** [comment]
If I understand you correctly, you want some additional features to the first version. 
* the burger icon to toggle the menu in the fixed line
* a floating button with a scroll to top function which is shown when the user scrolled to out of the menu

What exactly do you mean with "user control how to show"? should the user configure preferences which version they receive?

**Comment by romani** [comment]
> the burger icon to toggle the menu in the fixed line

no.
lets look on released version of website. I think we need to keep layout as is, BUT we will let user collapse 
![menu-with-collapse-expamnd](https://user-images.githubusercontent.com/812984/86608326-36c68680-bf5f-11ea-8f75-a99f9cc73b82.png)
in mobile mode it is collapsed by default. Collapsed mean that it will be seen as div and icon "expand".

**Comment by jo2** [comment]
Ok I think i have a clearer image of what you have in mind, now.
* Instead of the burger icon we make a small div containing one line "> Menu" (where > stands for the dropdown arrow) or "Menu x" (where x stands for the closing icon).
* On click on this div the menu expands to full size without collapsing functionality of the single menu items.
* On click on the Headline of the expanded menu or the x icon, the menu will collapse itself. 

Is that correct?

**Comment by romani** [comment]
yes, for some reason I wanted to collapse/expand to be on right side of scree to ease navigation by hand but it is not critical.
Please render us some preview site to be 100% sure we are on the same page on how it should behave and we can polish view later on.

**Comment by romani** [comment]
> a floating button with a scroll to top function which is shown when the user scrolled to out of the menu

I find this useful even for desktop website, as pages are long and to get back to menu(on top) you need to scroll A LOT.
if you can make it for all - it would be awesome.

**Comment by jo2** [comment]
I'll add the floating button for both desktop and mobile view in the PR for this issue if that is ok with you. If not, I'll create a new issue for that.

**Comment by jo2** [comment]
Meanwhile I uploaded a [WIP](https://jo2.github.io/checkstyle-menu/). In the next step I'd add a dropdown icon to the menu and an upward arrow to the to top button. But first please let me know what you think.

**Comment by romani** [comment]
>If not, I'll create a new issue for that.

Up to you. Such changes will not affect plugins and users, I am more relaxed on rules in this case.

**Comment by jo2** [comment]
For now I did it in the same code base so it'd be in the same PR.

**Comment by romani** [comment]
> But first please let me know what you think.

I found auto collapse of menu on scroll a bit unexpected. It is better to remove this feature. It does not allow to scroll menu and read items on end of menu.
All other looks good.

**Comment by jo2** [comment]
I can't reproduce auto collapse of menu on scrolling in neither chrome or firefox and tbh it shouldn't be there. What browser are you using?

**Comment by romani** [comment]
@jo2 ,

on Pixel 2 XL (Android 10, Build/QQ3A 200605.001), 
Chrome 83.0.4103.106,
the same behavior is on FF (android latest).

Video - https://drive.google.com/file/d/1X2BvtoMnKW4PJZaO3WYdMCd8DdvqAm4o/view?usp=sharing

**Comment by romani** [comment]
@jo2, if menu is problematic, please send PR for floating button to scroll to top.

**Comment by jo2** [comment]
I'll extract the to top button from this pr. I'm rather confused by this behaviour. I'll have to take a deeper look into it which may take some time.

**Comment by romani** [comment]
Did you reproduce auto collapse on your side?

**Comment by jo2** [comment]
Yeah, I managed to reproduce it. I assume it has to do something with mobile pointing and scrolling. if i scroll very slow on mobile, the menu stays open. The scrolling speed needed for that is about as fast as the scrolling speed I can reach using a simulated mobile device. But when I scoll faster, the menu collapses. I now have to analyze why the collapsing is triggered.

**Comment by romani** [comment]
@jo2 , let me know if there is some issues, we can do some changes in how it should behave or .....
I highly appreciate your help with website.

**Comment by jo2** [comment]
At the moment I sadly don't have that much time to invest into the in detail analysis of this behaviour. I hope it gets better mid next month. If you have any ideas or motivation to some analysis that is welcomed of course.

**Comment by romani** [comment]
@jo2 , please share a code that you did for  image with expand/collapse

**Comment by jo2** [comment]
https://github.com/jo2/checkstyle/tree/collapse-expand-menu here is the code i've written so far.

**Comment by Zopsss** [comment]
I'm working on this issue. I have gone through all the discussion in this issue and will try to provide a solution according to it.

**Comment by romani** [comment]
@jo2 , fyi, your 4 years old efforts are still alive and in high use!!!

@Zopsss , thanks a lot for fixing this issue.
